File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.ts
function letterCombinations(digits: string): string[] {
    if(!digits) return [];
    const all_combinations = [];
    let combination = "";
    const numberToLetterMapping = ["","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"];
    const recursivelyBacktrack=(currentDigit:string, currentCombination:string)=>{
        if(!currentDigit){
            all_combinations.push(currentCombination);
            return
        }
       for(let letter of numberToLetterMapping[ Number(currentDigit[0])]){
            let newCombination = currentCombination + letter;
            recursivelyBacktrack(currentDigit.slice(1),newCombination);
        }
    }
    recursivelyBacktrack(digits,combination);
    return all_combinations;
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        
        digit_to_letters = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        all_combinations = []
        
        def generate_combinations(digits_left, current_combination):
            if not digits_left:
                all_combinations.append(current_combination)
                return
            
            for letter in digit_to_letters[int(digits_left[0])]:
                generate_combinations(digits_left[1:], current_combination + letter)
        
        generate_combinations(digits, "")
        return all_combinations